llvm: Fix building llvm@4:9 using %clang@6: and %gcc@10:

bernhardkaindl opened the pull request:
@trws wrote here below
"... getting older LLVM (even 9) to build on newer Linux systems has proven to be a major problem in some cases because of things like this."

# This fixes building these older LLVM versions with more recent compilers:
## LLVM @8:11
Fix build of MicrosoftDemangle code (add missing include)
## LLVM@5:7
- Fix building sanitizers: https://github.com/llvm/llvm-project/commit/947f9692440836dcb8d88b74b69dd379d85974ce
## LLVM@4:5
 - `sys/ustat.h` has been removed in favour of statfs from glibc-2.28. Use fixed sizes (applied fix from llvm-6.0.1)
 -  Fix `lld` templates: https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=230463
 -  Add missing `include <cstddef>` for `std:size_t` in `llvm@4:5`,
    when built with '%clang@7:'

(and more fixes)

trws commented on var/spack/repos/builtin/packages/llvm/package.py:
```diff
@@ -282,6 +285,13 @@ class Llvm(CMakePackage, CudaPackage):
     # https://spack.readthedocs.io/en/latest/packaging_guide.html#making-a-package-discoverable-with-spack-external-find
     executables = ['clang', 'flang', 'ld.lld', 'lldb']
 
+    def patch(self):
```
This looks like it should have an @when decorator on it. What version of llvm/compiler-rt is this for?

bernhardkaindl replied:
Added and builds have succeeded.

bernhardkaindl committed:
llvm: Fix building llvm@4:9 using %clang@6: and %gcc@10:

haampie commented on var/spack/repos/builtin/packages/llvm/package.py:
```diff
@@ -282,6 +287,26 @@ class Llvm(CMakePackage, CudaPackage):
+        # This source defines some sanitizer values needing workarounds to build
+        # with newer compilers like gcc@10: and clang@13:
+        platform_posix = FileFilter(sanitizer + 'sanitizer_platform_limits_posix.cc')
+        if self.version >= Version('5.0.0') and self.version < Version('10.0.0'):
```
You could write this as `self.version in spack.version.ver('5:9')`

haampie commented:
I'm not sure if I'm a big fan of adding patches to support builds of particular llvm's with particular compilers. Generally I don't like `filter_file` cause it's very opaque. I'd rather have original llvm sources, and if a particular compiler is not supported, we add a conflict. Otherwise this package.py will be growing and growing and will be very hard to maintain.

trws commented:
> I'm not sure if I'm a big fan of adding patches to support builds of particular llvm's with particular compilers. Generally I don't like `filter_file` cause it's very opaque. I'd rather have original llvm sources, and if a particular compiler is not supported, we add a conflict. Otherwise this package.py will be growing and growing and will be very hard to maintain.

In general I agree, but at the same time getting older LLVM (even 9) to build on newer Linux systems has proven to be a major problem in some cases because of things like this.  As to the approach, how feasible would it be for this to just be a new copy of the patch file? If we can get away with that it would be a lot easier to parse what's going on.

bernhardkaindl commented:
> > I'm not sure if I'm a big fan of adding patches to support builds of particular llvm's with particular compilers. Generally I don't like `filter_file` cause it's very opaque. I'd rather have original llvm sources, and if a particular compiler is not supported, we add a conflict. Otherwise this package.py will be growing and growing and will be very hard to maintain.
> 
> In general I agree, but at the same time getting older LLVM (even 9) to build on newer Linux systems has proven to be a major problem in some cases because of things like this. As to the approach, how feasible would it be for this to just be a new copy of the patch file? If we can get away with that it would be a lot easier to parse what's going on.

Yes, my initial push of the PR was a quick hack which only covered a few issues and I agree it was very opaque to review and not documented at all. I've taken your points, made patches and added references and/or descriptions to all the patches to keep the recipe tidy and well readable.

On the conflicts, I'm siding with @trws here: conflicts prevent to build these older versions:

* I now replaced all conflicts and filter_file() calls that I added with proper clean patches where possible.
* Where not possible, the remaining conflicts now target specific variants which are possibly not feasible to fix. There are more conflicts than I'd like but it's better to let clingo disable a variant thru a conflict than to abort the build later.

This fixes the build of LLVM `@4:9` with `%clang@6:` and `%gcc@10:` and did countless test builds to verify it all works in all the major combinations of `%gcc`/`%clang@versions` with `+compiler-rt` and `+lld` and without them added to check the results.

In any case, this is a incredibly huge step forward, fixing/enabling the build of older llvm versions in any combination which was feasible to fix within the available time.

I think the amount of patches and all the references speak for themself:
This is now fully done and ready to be merged.

trws commented on var/spack/repos/builtin/packages/llvm/package.py:
```diff
@@ -167,7 +167,8 @@ class Llvm(CMakePackage, CudaPackage):
     depends_on("python", when="@5:+python")
-    depends_on("z3", when="@9:")
+    # While not required for @8:9, cmake can find a z3 on the system, causing failures:
+    depends_on('z3', when='@8:')
```
Could you add one thing here, since z3 is now on the required list back to 8, add the cmake define to use it: `LLVM_ENABLE_Z3_SOLVER`.  In principle it should be on if found, but if we're requiring it, I'd rather have it in the build so we find out if something goes wrong with the dependency.

bernhardkaindl replied:
@trws Done with update: It is documented @ https://llvm.org/docs/CMake.html to be an option for the clang analyzer
Thus I turned it into a variant for people using the clang static analyzer and need it.

Two smaller fixes added as well:
* The `no_cyclades.patch` backported was included in `12.0.1`, so `12.0.0` still needs it.
* The backport of the ipc_perm patch for the sanitizer is also needed on Ubuntu `focal` with `gcc@9.3.0`

trws reviewed (changes requested):
I'm a whole lot happier with this, thank you for going through this so thoroughly.  Would you mind adding the Z3 define? With that, I'm good to merge.

Commit:
Add z3 variant, fix @:9%gcc@9: w/ glibc2.31, fix no_cyclades range

bernhardkaindl commented:
@trws: I added the only remaining change that you demanded using a variant because Z3 is only used for the Clang static analyzer...

With this, you said, you are good to merge.

Further fixes:
- Fix install `@:9%gcc@9:` with glibc-2.31
- Minor fix for the version range of the no_cyclades patch.
- Fix install `clang@12:13%clang@13` running on Ubuntu-18.04 with `clang-13` installed in `/usr/bin`
- Added the missing newer cmake dependency since LLVM-12:
```
depends_on('cmake@3.13.4:', type='build', when='@12:')
```

dylex commented:
Since this I'm getting an error building llvm 10 or 11 on develop:
```
>spack patch llvm@10.0.1
Reversed (or previously applied) patch detected!  Assume -R? [n]
Apply anyway? [n]
1 out of 1 hunk ignored -- saving rejects to file llvm/include/llvm/Demangle/MicrosoftDemangleNodes.h.rej
==> Patch /mnt/home/dylan/spack/var/spack/repos/builtin/packages/llvm/missing-includes.patch failed.
==> Error: Command exited with status 1:
'/bin/patch' '-s' '-p' '1' '-i' '/mnt/home/dylan/spack/var/spack/repos/builtin/packages/llvm/missing-includes.patch' '-d' '.'
```
is this just me? Looking at the llvm 10.0.1 source and this patch, it does seem like this patch does not apply there.

trws commented:
This change shouldn't have impacted patches to llvm 10.x, only 12 and 4:9.  I see the missing includes patch in there though, was that intentional or did you mean to hit through the 9 series @bernhardkaindl?

bernhardkaindl commented:
@trws I'm on vacation and mobile only at the moment and can't look, so I can only say what I remember:

While working on this PR, the change where failing to patch silently didnt fail the build, so I didn't notice the patch error. 

I'm guessing the patch should only apply for a range of older releases, it seems like before 10, so maybe only for @8:9.
